### PR TITLE
electron: update rendering of menu items

### PR DESCRIPTION
**Pull request by vince-fugnitto**
#### What it does

The following pull-request updates the rendering of **electron context-menu items**  based on a command's `isEnabled`. In electron, menu items cannot be disabled (grayed-out) like the browser so we must rely on hiding elements instead of disabling them. With this change, menu items in electron are not rendered if their `isEnabled` is not true.

For example, language-server commands (monaco) are always populated in electron menus which does not make sense since such commands (ex: 'go to definition') do not make sense when they are no-op. Omitting the rendering of these items also aligns with the menu behavior visible from vscode.

_Master_:

<div align='center'>

Screen Shot 2020-05-22 at 8 51 10 PM

</div>

_Pull Request_:

<div align='center'>

Screen Shot 2020-05-22 at 9 34 04 PM

</div>

#### How to test

1. execute the command `Reset Workbench Layout`
2. verify that the main-menu successfully displays the menu items (https://github.com/eclipse-theia/theia/pull/7869#issuecomment-633519842)
3. open a typescript file and trigger the context-menu 
(language feature menu items should be available)
4. open a markdown, image, textfile and trigger the context-menu 
(language feature menu items should not be available)

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

Signed-off-by: vince-fugnitto <vincent.fugnitto@ericsson.com>

### CODE REVIEW

**Comment by kittaakos** [comment]
Two remarks:
 - VS Code always shows application menu items even if they are not enabled. I assume, if we do this for the context menu, we will do the same for the app menu; do we want to diverge from VS Code's behavior?
 - If we decide to hide not enabled items, we have to adjust [this](https://github.com/eclipse-theia/theia/pull/7869/files#diff-989f20441ef194c1d93996cee1d24245L190-L191) code too.

**Comment by kittaakos** [comment]
There is another side-effect of this PR, we do not show non-enabled menu items on start, so for instance, the `Selection` menu will contain only `Select All`. Since we [do not support dynamic menu items](https://github.com/eclipse-theia/theia/issues/446) in electron, the `Selection` menu content won't ever get updated unless the window has a `blur` then `focus` event:

Screen Shot 2020-05-25 at 13 06 50

Screen Shot 2020-05-25 at 13 10 43


I tend to not accept this PR with the current state. I prefer to see the menu items even if they're NOOP than not seeing them.

**Comment by vince-fugnitto** [comment]
Thank you for your feedback @kittaakos!
> 
> * VS Code always shows application menu items even if they are not enabled. I assume, if we do this for the context menu, we will do the same for the app menu; do we want to diverge from VS Code's behavior?

We will need to find a solution which does not affect the main-menu, but rather only the context-menu (due to the dynamic menu issue). In VS Code, the context-menu only displays enabled menu items, it will not display no-op menu items (similarly to this pull-request).

> * If we decide to hide not enabled items, we have to adjust [this](https://github.com/eclipse-theia/theia/blob/f9d05b15ab9201b0198226807f89f0fffd908e49/packages/core/src/electron-browser/menu/electron-main-menu-factory.ts#L189-L192) code too.

The idea is to hide context-menu items only.

> I tend to not accept this PR with the current state. I prefer to see the menu items even if they're NOOP than not seeing them.

I agree, I'll need to dive deeper to see if it's possible only to modify the context-menu and not the main-menu. 

> There is another side-effect of this PR, we do not show non-enabled menu items on start, so for instance, the Selection menu will contain only Select All.

I wasn't able to reproduce this issue with my changes, do you have the steps to reproduce?

**Comment by kittaakos** [comment]
> do you have the steps to reproduce?

You can try resetting the workbench layout.

**Comment by kittaakos** [comment]
> In VS Code, the context-menu only displays enabled menu items

> The idea is to hide context-menu items only.

Do you want to align the browser behavior with electron? AFAIK, we disable but show the context menu items in browser env.

**Comment by vince-fugnitto** [comment]
> > In VS Code, the context-menu only displays enabled menu items
> 
> > The idea is to hide context-menu items only.
> 
> Do you want to align the browser behavior with electron? AFAIK, we disable but show the context menu items in browser env.

@kittaakos I was only planning on updating the **electron** target to hide disabled menu items (similarly to vscode) and not update the browser which does not have this limitation. In the browser, disabled menu items are properly decorated (grayed-out).

**Comment by kittaakos** [comment]
> @kittaakos I was only planning on updating the **electron** target to hide disabled menu items (similarly to vscode) and not update the browser which does not have this limitation. In the browser, disabled menu items are properly decorated (grayed-out).

👍 I am fine with it, I just wanted to know what to verify once it's ready.

**Comment by vince-fugnitto** [comment]
@kittaakos I modified the code to only affect context-menus and not the main-menu.
The behavior is now aligned with vscode (hiding disabled menu items from the context-menu).

File: packages/core/src/electron-browser/menu/electron-main-menu-factory.ts [hunk 1]
```diff
@@ -136,6 +136,11 @@ export class ElectronMainMenuFactory {
                     continue;
                 }
 
+                // We should omit rendering context-menu items which are disabled.
+                if (!this.commandRegistry.isEnabled(commandId, ...args) && !!isContextMenu) {
```
Comment (kittaakos) [inline packages/core/src/electron-browser/menu/electron-main-menu-factory.ts:140]: [minor]: Can you swap the conditions? No need to check `isEnabled` when `isContextMenu` is falsy.

File: packages/core/src/electron-browser/menu/electron-main-menu-factory.ts [hunk 2]
```diff
@@ -76,14 +76,14 @@ export class ElectronMainMenuFactory {
...
-
+        const template = this.fillMenuTemplate([], menuModel, args, true);
         return electron.remote.Menu.buildFromTemplate(template);
     }
 
     protected fillMenuTemplate(items: Electron.MenuItemConstructorOptions[],
         menuModel: CompositeMenuNode,
-        args: any[] = []
+        args: any[] = [],
+        isContextMenu?: boolean,
```
Comment (akosyakov) [inline packages/core/src/electron-browser/menu/electron-main-menu-factory.ts:86]: (nit) for last optional it is better to use an object, it is easy to extend without breaking later

**Comment by vince-fugnitto** [comment]
@kittaakos @akosyakov thank you for the feedback, I've updated and retested the code based on your feedback :)

**Pull request merged**